lib/cc/cli/version_checker.rb

```diff
@@ -0,0 +1,69 @@
+require "uuid"
+
+module CC
+  module CLI
+    class VersionChecker
+      include CC::CLI::Output
+
+      VERSION_CHECK_TIMEOUT = 60 * 60 # 1 Hour in seconds
+      DEFAULT_VERSIONS_URL = "https://versions.codeclimate.com".freeze
+
+      def check
+        print_new_version_message if outdated?
+      rescue => error
+        CLI.debug(error)
+      end
+
+      private
+
+      def outdated?
+        api_response["outdated"] == true
+      end
+
+      def latest_version
+        api_response["latest"]
+      end
+
+      def print_new_version_message
+        warn "A new version (v#{latest_version}) is available"
+      end
+
+      def api_response
+        @api_response ||=
+          begin
+            JSON.parse(api_response_body)
+          rescue JSON::ParserError => error
+            CLI.debug(error)
+            # We don't know so pretend all is peachy. We'll try again next time.
+            {
+              "latest" => version,
+              "outdated" => false,
+            }
+          end
+      end
+
+      def api_response_body
+        if http_response.is_a? Net::HTTPSuccess
+          http_response.body
+        else
+          raise Net::HTTPFatalError.new("HTTP Error", http_response)
+        end
+      end
+
+      def http_response
+        @http_response ||=
```

Comment:
Oh I see. Sorry I misread that, maybe that's an indication that the methods could be broken up a bit differently and be easier to follow? I'm not sure, but FWIW I might've written this this way:

```rb
# Not in love with the names, just trying to show division of responsibility...

def remote_version_info
  # only memoize one thing, the final result
  @remote_version_info ||= get_remote_version_info
end

def get_remote_version_info
  # response parsing, not memoized
  response = http_response

  if response.is_a?(...)  
    JSON.parse(response.body)
  else
    raise ...
  end
rescue JSON::ParserError
  { ... }
end

def http_response
  # Net::HTTP stuff, not memoized
end
```

This is really similar to what you have now, it just avoids memoizing multiple intermediate results which can be confusing and hide bugs (in my opinion and experience). This also would address two explicit `begin`s, which I do consider smells.